--- ./cleanfeed.org	2008-08-12 09:49:59.932847792 +0000
+++ ./cleanfeed	2008-08-12 09:50:48.423467334 +0000
@@ -15,7 +15,7 @@
 # Directory where cleanfeed.local and the other configuration files live.
 # Set this to undef to not use any external file.
 
-$config_dir = '/usr/local/news/cleanfeed/etc';
+$config_dir = '/etc/news';
 
 ##############################################################################
 # Server configuration
--- cleanfeed/etc/cleanfeed.local.orig	2013-06-19 21:39:58.076002907 +0200
+++ cleanfeed/etc/cleanfeed.local	2013-06-19 21:41:49.602667248 +0200
@@ -156,15 +156,15 @@
     %config_local = (
         block_late_cancels => 1,
         block_user_cancels => 1,
-        active_file => '/usr/local/news/db/active',
-        statfile => '/usr/local/news/cleanfeed/cleanfeed.stats',
-#       html_statfile => '/news/log/cleanfeed.stats.html',
+        active_file => '/var/lib/news/active',
+        statfile => '/var/log/news/cleanfeed.stats',
+#       html_statfile => '/var/log/news/cleanfeed.stats.html',
         stats_interval => 300,
         do_scoring_filter => 0,
         do_emp_dump => 1,
-        emp_dump_file => '/usr/local/news/tmp/empdump',
+        emp_dump_file => '/var/lib/news/tmp/empdump',
         md5_skips_followups => 0,
-        debug_batch_directory => '/usr/local/news/cleanfeed/log',
+        debug_batch_directory => '/var/lib/news/cleanfeed/log',
         flood_groups => '^sci\.crypt$|^news\.software\.nntp$|^alt\.free\.newsservers',
         MD5RateCutoff => 3,
         PHNRateBaseInterval => 1800,
